src/python/pants/backend/terraform/target_types.py

```diff
@@ -51,5 +57,71 @@ class TerraformModuleTarget(Target):
     )
 
 
+class TerraformBackendConfigField(OptionalSingleSourceField):
+    alias = "backend_config"
+    help = "Configuration to be merged with what is in the configuration file's 'backend' block"
+
+    def empty(self) -> TerraformBackendConfigField:
+        """Clear the backend config to ensure that."""
+        return TerraformBackendConfigField(None, self.address)
+
+
+class TerraformVarFilesField(MultipleSourcesField):
+    alias = "var_files"
+    expected_file_extensions = (".tfvars",)
+    help = generate_multiple_sources_field_help_message(
+        "Example: `var_files=['common.tfvars', 'prod.tfvars']`"
+    )
+
+
+class TerraformExtraArgs(StringSequenceField):
+    alias = "extra_args"
+    help = help_text(
+        """
+        Extra arguments for `terraform apply`
+        Example: `extra_args=["-var" "'var0=hihello'"]`"
+        """
+    )
+
+
+class TerraformDeploymentTarget(Target):
```

Comment:
Does this need to be a separate target from `terraform_module`? Why not have the extra fields live as `deploy_FOO` fields on a `terraform_module`?

Reply:
And then the deploy implementation can error if the `terraform_module` is not in fact a deployable Terraform module.

Reply:
I think there isn't a technical need for a separate target type. Terraform doesn't separately model root modules and shared modules. It's technically possible to try to treat them all the same, and try running `init` and `apply`. There are cases where that doesn't work, there's a bit more in the linked issue.

I think making all modules targettable as deployments has consequences for Pants. It means that every module will generate its own resolve and will be targetted by `generate-lockfiles` and `export` (future plans). That might not be very useful if there's one root deployment which has several internal modules. [ref](https://pantsbuild.slack.com/archives/C0D7TNJHL/p1684817389109879) (I think we could make resolves which include multiple modules by implementing logic to solve the requirements and search for the modules, but that's a lot of implementation on our part.)

I personally think it makes sense to have a separate target type for deployments. This lets us have multiple deployments which reference the same Terraform module but with different backend configs and variables. Something like [this layout](https://www.codurance.com/publications/2020/04/28/terraform-with-multiple-environments#Using-Folders-Method-2) would let us have a build file something like
```python
terraform_module()
terraform_deployment(name="DEV", vars_file="dev/terraform.tfvars")
terraform_deployment(name="TEST", vars_file="test/terraform.tfvars")
terraform_deployment(name="PROD", vars_file="prod/terraform.tfvars")
```
It also lets me show my team members what's meant as a deployable root module and what's meant to be a shared module.

Reply:
I also like having a separate target because it allows to run `pants experimental-deploy ::` (or any other kind of wildcard) without making Pants fail when it tries to do so on a non-deployable module.